samples/coop_synchronization/src/main.c

```diff
@@ -0,0 +1,142 @@
+/*
+ * Copyright (c) 2012-2014 Wind River Systems, Inc.
+ * Copyright (c) 2023 Advanced Micro Devices, Inc. (AMD)
+ * Copyright (c) 2023 Alp Sayin <alpsayin@gmail.com>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/kernel.h>
+#include <zephyr/sys/printk.h>
+
+/*
+ * The coop synchronization demo has two threads that utilize semaphores and k_yield
+ * to take turns printing a greeting message with a somewhat uncontrolled rate.
+ * The rate is controlled via a busy waiting loop. The advantage of this demo is that
+ * it'll work even if there are no timing utilities provided by the architecture.
+ */
+
+#define PIN_THREADS (IS_ENABLED(CONFIG_SMP)		  \
+		     && IS_ENABLED(CONFIG_SCHED_CPU_MASK) \
+		     && (CONFIG_MP_NUM_CPUS > 1))
+
+/* size of stack area used by each thread */
+#define STACKSIZE 1024
+
+/* scheduling priority used by each thread */
+#define PRIORITY 7
+
+/*
+ * @param my_name      thread identification string
+ * @param my_sem       thread's own semaphore
+ * @param other_sem    other thread's semaphore
+ */
+void helloLoop(const char *my_name, struct k_sem *my_sem,
+	struct k_sem *other_sem, int incrementer)
+{
+	const char *tname;
+	uint8_t cpu;
+	struct k_thread *current_thread;
+	int task_local_counter = 0;
+
+	while (1) {
+		/* take my semaphore */
+		k_sem_take(my_sem, K_FOREVER);
+
+		int key1 = irq_lock();
```

Comment:
irq_lock() is heavily disfavored in any code that can run in SMP context, where it is a global lock (and a very expensive one).  Use a spinlock here; which AFAICT is only needed because you want to check the current CPU and without it you might migrate?

Reply:
It was more of my way of making things more "interesting" and testing things are still OK when yielding with `irq_lock`. Very specifically I was trying to see interrupt-enable in microblaze machine-status-reg was restored properly.
I had not thought about SMP context. Your concern is a valid one but would not really affect the sample. Migration is not an issue as long as the thread context is restored correctly.
Also, -I might be very off about this- but wouldn't the spinlock spin forever without the timer to kick-in and change context ?

p.s. I might be off about a lot of things please bear with me 🙃

Reply:
> It was more of my way of making things more "interesting" and testing things are still OK when yielding with irq_lock

As we still take irq_lock inside of spinlock - I'd say that it's safe to simply replace irq_lock with spilock here. You still will be yelding with irqs locked.

Reply:
> evgen

Done. Thanks for the idea.